docs/design/pr-builder.rst

```diff
@@ -0,0 +1,146 @@
+We have to add ``QuerySet`` for PR versions also.
+
+Excluding PR Versions from Elasticsearch Indexing
+-------------------------------------------------
+
+We should exclude to PR Versions from being Indexed to Elasticsearch.
+We need to update the queryset to exclude PR Versions.
+
+Creating Versions for Pull Requests
+-----------------------------------
+
+If the Github webhook event is ``pull_request`` and action is ``opened``,
+it means a pull request was opened in the projects repository.
+We can create a ``Version`` from the Payload data and trigger a initial build for the version.
+A version will be created whenever RTD receives an event like this.
+
+Triggering Build for New Commits in a Pull Request
+--------------------------------------------------
+
+We might want to trigger a new build for the PR version if there is a new commit on the PR.
+
+Status Reporting to Github
+--------------------------
+
+We could send build status reports to Github. We could send if the build was Successful or Failed.
+We can also send the build URL. By this we could show if the build passed or failed on Github something like travis-ci does.
+
+As we already have the ``repo:status`` scope on our OAuth App,
+we can send the status report to Github using the Github Status API.
+
+Sending the status report would be something like this:
+
+.. http:post:: /repos/:owner/:repo/statuses/:sha
+
+.. code:: json
+
+   {
+       "state": "success",
+       "target_url": "<pr_build_url>",
+       "description": "The build succeeded!",
+       "context": "continuous-documentation/read-the-docs"
+   }
+
+Storing Pull Request Docs
+-------------------------
+
+We need to think about how and where to store data after a PR Version build is finished.
+We can store the data in a blob storage.
+
+Deleting a PR Version
+---------------------
+
+We can delete a PR version when:
+
+- A pull request is ``closed``. Github Webhook event (Action: ``closed``, Merged: ``False``)
+- A pull request is ``merged``. Github Webhook event (Action: ``closed``, Merged: ``True``)
+- A PR Version has reached its life time
+
+We might want to set a life time of a PR version in case:
+
+- We don't receive webhook event for a pull request being closed/merged
+- If a PR is stale (not merged for a long time).
+
+We need to delete the PR Version alongside the Build data from the blob storage.
+
+Excluding PR Versions from Search Engines
+-----------------------------------------
+
+We should Exclude the PR Versions from Search Engines,
+because it might cause problems for RTD users.
+As users might land to a pull request doc but not the original Project Docs.
+This will cause confusion for the users.
+
+Serving PR Docs
+---------------
+
+We need to think about how we want to serve the PR Docs.
+
+- We could serve the PR Docs from another Domain.
+- We could serve the PR Docs using ``<pr_number>`` namespace on the same Domain.
+  ``https://<project_slug>.readthedocs.io/en/pr/<pr_number>/``
```

Comment:
I think that if we keep our current URLs it could work. Example, https://<project_slug>.readthedocs.io/<language_code>/<version_slug>/

This way, we will stick to the `language_code` defined in the `conf.py` (`ja`, e.g) of that PR and we will can force the `Version.slug` to be `pr5726` when a version is created. The resulting URL would be https://myproject.readthedocs.io/ja/pr5726/. This will interferes with branches and tags named in the same way, though.

On the other hand, single version project URLs will be https://myproject.readthedocs.io/pr5726/, which will interferes with any folder named `pr5726` in the official project docs.

Another idea could be to use https://pr.<project_slug>.readthedocs.io/<pr_number>/.

I suppose that using any other approach that does not match our current mapping or adds "an special case inside our current mapping" will be complicated to adapt around all the code we have already written.

Reply:
i like this one more `https://myproject.readthedocs.io/ja/pr5726/`. can we make the slug like this `pr-5726`
Both of them are good ideas. I'll add them in the Docs :+1:

Reply:
In the second case there is no reason to add the language code in the URL from my point of view.

In the first case, it's useful to keep compatibility with the rest of URL handling and just base the PR serving on the slug being `pr-<number>`.